### ISSUE TITLE
api.markVariableAsUsed does not check Node.js scope

### ISSUE THREAD

**Issue by yannickcr** [eslint/eslint#2089]
The Node.js scope is not checked by the api.markVariableAsUsed function making these variables impossible to mark as used.

```javascript
"use strict";

var eslint = require("./lib/api.js").linter;

var code = "var a = 1, b = 2;";

eslint.reset();
eslint.on("Program:exit", function() {
    eslint.markVariableAsUsed("a"); // return false, the variable was not found in the scope
});

eslint.verify(code, { env: { node: true }}, "test.js", true);
```

**Comment by nzakas** [comment]
I'm not sure I'd call this a bug with the method, it seems more a bug with how it's being used. The method searches from the current scope up to the global scope. If you're calling it in "Program", then you're saying "only search the global scope".

However, in almost all cases you should not be calling `markVariableAsUsed()` in the global scope, you should be calling it where the identifier is actually used. 

Can you show me how you're using it?

**Comment by nzakas** [comment]
Okay, I saw your other bug and I get it now. While I'd still recommend not using `markVariableAsUsed()` in the global scope, rather use it in the appropriate JSX element context, I can see how this would be confusing.

**Comment by yannickcr** [comment]
I'm using it to mark as used the variables in the JSX code.

You can see the rule here: https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-uses-vars.js

**Comment by nzakas** [comment]
That looks like it should work (since you're not in the global scope). Are you still seeing an error?

**Comment by yannickcr** [comment]
Yes, there is an error when the variable is used in the global scope.

You can see some example of problematic code [here](https://github.com/yannickcr/eslint-plugin-react/issues/19) and [here](https://github.com/yannickcr/eslint-plugin-react/issues/18#issuecomment-80762115)

**Comment by nzakas** [comment]
Hmm, I'm I'm looking at this correctly, I'm not sure `markVariableAsUsed()` is the problem. Because it searches from the current scope up to the global scope, it should still find any variable that was defined in the Node.js scope. It seems more likely the problem is in `no-undef` and `no-unused-vars` not checking the Node.js scope.

**Comment by yannickcr** [comment]
Sorry but I do not see the relation with `no-undef` or `no-unused-vars`

To retake the above example:
```javascript
"use strict";

var eslint = require("./lib/api.js").linter;

var code = "var a = 1, b = 2;";

eslint.reset();
eslint.on("Program:exit", function() {
    eslint.markVariableAsUsed("a"); // return false, the variable was not found in the scope
});

eslint.verify(code, { env: { node: true }}, "test.js", true);
```

But if we are not in the the Node.js environment:
```javascript
"use strict";

var eslint = require("./lib/api.js").linter;

var code = "var a = 1, b = 2;";

eslint.reset();
eslint.on("Program:exit", function() {
    eslint.markVariableAsUsed("a"); // return true, the variable was found in the scope
});

eslint.verify(code, {}, "test.js", true);
```

When in a Node.js scope, the variable `a` and `b` cannot be found in scope.variables but are in scope.childScopes[0].variables which is not currently checked in the actual `markVariableAsUsed()`. Seems to be a bug with `markVariableAsUsed()` to me :)

Pull request #2090 should fix this.

**Comment by nzakas** [comment]
What I meant was the original issue you reported had nothing to do with this. This is a separate issue.

**Comment by nzakas** [comment]
Working on this.
